src/com/seafile/seadroid2/gesturelock/DefaultAppLock.java

```diff
@@ -0,0 +1,126 @@
+ */
+package com.seafile.seadroid2.gesturelock;
+
+import android.app.Activity;
+import android.app.Application;
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+import com.seafile.seadroid2.SettingsManager;
+import com.seafile.seadroid2.ui.activity.UnlockGesturePasswordActivity;
+
+import java.util.Arrays;
+import java.util.Date;
+
+/**
+ * Implementation of AppLock
+ */
+public class DefaultAppLock extends AbstractAppLock {
+    public static final String DEBUG_TAG = "DefaultAppLock";
+
+    private Application currentApp; //Keep a reference to the app that invoked the locker
+    private String lastPausedActivityName;
+    private String currentResumeActivityName;
+
+    public DefaultAppLock(Application currentApp) {
+        super();
+        this.currentApp = currentApp;
+    }
+
+    public void enable() {
+        if (android.os.Build.VERSION.SDK_INT < 14)
+            return;
+
+        if (isPasswordLocked()) {
+            currentApp.unregisterActivityLifecycleCallbacks(this);
+            currentApp.registerActivityLifecycleCallbacks(this);
+        }
+    }
+
+    @Override
+    public void disable() {
+        if (android.os.Build.VERSION.SDK_INT < 14)
+            return;
+
+        currentApp.unregisterActivityLifecycleCallbacks(this);
+    }
+
+    //Check if we need to show the lock screen at startup
+    public boolean isPasswordLocked() {
+
+        return SettingsManager.instance().isGestureLockEnabled();
+    }
+
+    @Override
+    public void onActivityPaused(Activity activity) {
+        Log.d(DEBUG_TAG, "onActivityPaused");
+
+        if (activity.getClass() == UnlockGesturePasswordActivity.class)
+            return;
+
+        lastPausedActivityName = activity.getClass().getName();
+
+        if ((this.appLockDisabledActivities != null) &&
+                Arrays.asList(appLockDisabledActivities).contains(activity.getClass().getName()))
+            return;
+    }
+
+    @Override
+    public void onActivityResumed(Activity activity) {
+        Log.d(DEBUG_TAG, "onActivityResumed");
+        if (activity.getClass() == UnlockGesturePasswordActivity.class)
+            return;
+
+        if ((this.appLockDisabledActivities != null) &&
+                Arrays.asList(appLockDisabledActivities).contains(activity.getClass().getName()))
+            return;
+
+        currentResumeActivityName = activity.getClass().getName();
+
+        if (mustShowUnlockSceen()) {
+            Intent i = new Intent(activity.getApplicationContext(), UnlockGesturePasswordActivity.class);
+            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            activity.getApplication().startActivity(i);
+            return;
+        }
+
+    }
+
+    private boolean mustShowUnlockSceen() {
+        // don`t show unlock screen when navigating inside of App
+        if (currentResumeActivityName != null && !currentResumeActivityName.equals(lastPausedActivityName)) {
```

Comment:
This logic is not correct. I can get around the gesture lock protection by this:

- open seadroid from home screen, input the correct gesture lock
- go back to home screen
- wait for 5 minutes to let the lock expire
- Now open gallery app, view a picture, click the "share button", and choose "share to seafile"
- seafile would not ask for the gesture lock, which it should do